nix/flake/flakeref.go

```diff
@@ -156,24 +166,64 @@ func parseURLRef(ref string) (parsed Ref, fragment string, err error) {
 		} else {
 			parsed.Path = refURL.Path
 		}
+
+		parsed.NARHash = refURL.Query().Get("narHash")
+		parsed.LastModified, err = atoiOmitZero(refURL.Query().Get("lastModified"))
+		if err != nil {
+			return Ref{}, "", redact.Errorf("parse flake reference URL query parameter: lastModified=%s: %v", redact.Safe(parsed.LastModified), redact.Safe(err))
+		}
 	case "http", "https", "file":
 		if isArchive(refURL.Path) {
 			parsed.Type = TypeTarball
 		} else {
 			parsed.Type = TypeFile
 		}
-		parsed.Dir = refURL.Query().Get("dir")
+		query := refURL.Query()
+		parsed.Dir = query.Get("dir")
+		parsed.NARHash = refURL.Query().Get("narHash")
+		parsed.LastModified, err = atoiOmitZero(refURL.Query().Get("lastModified"))
+		if err != nil {
+			return Ref{}, "", redact.Errorf("parse flake reference URL query parameter: lastModified=%s: %v", redact.Safe(parsed.LastModified), redact.Safe(err))
+		}
+
+		// lastModified and narHash get stripped from the query
+		// parameters, but dir stays.
+		query.Del("lastModified")
+		query.Del("narHash")
+		refURL.RawQuery = query.Encode()
 		parsed.URL = refURL.String()
 	case "tarball+http", "tarball+https", "tarball+file":
 		parsed.Type = TypeTarball
-		parsed.Dir = refURL.Query().Get("dir")
+		query := refURL.Query()
+		parsed.Dir = query.Get("dir")
+		parsed.NARHash = refURL.Query().Get("narHash")
+		parsed.LastModified, err = atoiOmitZero(refURL.Query().Get("lastModified"))
+		if err != nil {
+			return Ref{}, "", redact.Errorf("parse flake reference URL query parameter: lastModified=%s: %v", redact.Safe(parsed.LastModified), redact.Safe(err))
+		}
 
+		// lastModified and narHash get stripped from the query
+		// parameters, but dir stays.
+		query.Del("lastModified")
+		query.Del("narHash")
```

Comment:
seems like these lines are repeated in many case-bodies. Pull into a function and call that function instead?

Reply:
Are you referring to the `query.Del` calls? I dunno, it seems like overkill to make those two lines a separate function.